working/control-flow-collections/feature-specification.md

```diff
@@ -0,0 +1,913 @@
+            `variable` for it. Otherwise, use the existing `variable` it refers
+            to.
+
+        1.  Evaluate `iterator.current` and bind it to `variable`.
+
+        1.  Evaluate the body element using this procedure in the scope of
+            `variable`.
+
+    1.  If the `for-in` element declares a variable, discard it.
+
+1.  Else, if `element` is a C-style `for` element:
+
+    1.  Evaluate the initializer clause of the element, if there is one.
+
+    1.  Loop:
+
+        1.  Evaluate the condition expression to a value `condition`. If there
+            is no condition expression, use `true`.
+
+        1.  Subject `condition` to boolean conversion to a value `result`.
+
+        1.  If `result` is not `true`, exit the loop.
+
+        1.  Evaluate the body element using this procedure in the scope of the
+            variable declared by the initializer clause if there is one.
+
+        1.  If there is an increment clause, execute it.
+
+1.  Else, if `element` is a spread element, see the relevant proposal.
+
+1.  Else, if `element` is a normal list element:
+
+    1.  Evaluate the element's expression to a value `value`.
+
+    1.  Append `value` to `list`.
+
+1.  Else, `element` has form `keyExpression: valueExpression`:
+
+    1.  Evaluate `keyExpression` to a value `key`.
+
+    1.  Evaluate `valueExpression` to a value `value`.
+
+    1.  Call `map[key] = value`.
+
+## Migration
+
+This is a non-breaking change that purely makes existing semantics more easily
+expressible, so there is no required migration.
+
+### Automated fixes
+
+It may be possible for tooling to detect some of the existing idioms that are
+better expressed using this new syntax and give the user the option to
+automatically change it to the new style. Cases where conditional logic has been
+hoisted all the way out of a collection may be hard to detect since the
+resulting code is pretty imperative.
+
+It should be possible to detect cases like:
+
+*   **Switching out an element using a conditional operator:**
+
+    ```dart
+    [
+      before,
+      condition ? first : second,
+      after
+    ]
+    ```
+
+    Fix:
+
+    ```dart
+    [
+      before,
+      if (condition) first else second,
+      after
+    ]
+    ```
+
+*   **Omitting an element using a conditional operator and `null` filtering:**
+
+    ```dart
+    [
+      before,
+      condition ? first : null,
+      after
+    ].where((e) => e != null).toList()
+    ```
+
+    Fix:
+
+    ```dart
+    [
+      before,
+      if (condition) first,
+      after
+    ]
+    ```
+
+*   **Using `Map.fromIterable()`:**
+
+    ```dart
+    Map.fromIterable(things,
+      key: (e) => someExpression(e),
+      value: (e) => anotherExpression(e)
+    )
+    ```
+
+    Fix:
+
+    ```dart
+    {
+      for (var e in things)
+        someExpression(e): anotherExpression(e)
+    }
+    ```
+
+    This may fall down if the two closures are complex but I think that's rare
+    in practice.
```

Comment:
(or the fix can name the functions first and use them in the map comprehension)

Reply:
True, but I think that would make the code *worse* in most users eyes.